### PR TITLE
switch to whitelist for corsMode

### PR DESCRIPTION
**Pull request by igrigorik**
Followup to https://github.com/w3c/beacon/pull/33#issuecomment-229896895. 

@annevk @toddreifsteck ptal.

### CODE REVIEW

**Comment by annevk** [comment]
Didn't you want to match `XMLHttpRequest` and `fetch()` by also allowing `ArrayBuffer` and friends? I wonder if they also allow some variant of `Blob` as long as the MIME type matches one of the allowed ones...

**Comment by toddreifsteck** [comment]
Yes, ArrayBuffer should be part of the whitelist. Blob will be processed by the no-cors algorithm within Fetch... The only downside to doing it this way is that it will not be redirectable like other input types.

**Comment by igrigorik** [comment]
@toddreifsteck @annevk good catch, added ArrayBuffer to the "no-cors" clause. Anything else?

**Comment by annevk** [comment]
You didn't cover the "and friends" part. I guess I should take some time to review this part of Fetch again, it's apparently a little flaky.

**Comment by annevk** [comment]
Yeah, it seems like Fetch has a bug. It sets the `Content-Type` header after the header check. 😟

**Comment by annevk** [comment]
Oh no, it's fine. It uses an append operation that's guarded.

**Comment by annevk** [comment]
So yeah, `fetch()` allows a `Blob` as long as its MIME type is safelisted.

**Comment by igrigorik** [comment]
@annevk which other friends are there? 

Re, "MIME type is safelisted" are you referring to [reading the response](https://fetch.spec.whatwg.org/#should-response-to-request-be-blocked-due-to-mime-type?)? Or, is there an equivalent check (which I can't find) for when request is made?

**Comment by annevk** [comment]
@igrigorik Uint8Array and such?

No, I'm referring to the value of the `Content-Type` header that the MIME type ends up in. And whether that header/value pair is safelisted or not.

**Comment by igrigorik** [comment]
@annevk ah, in which case [BufferSource](https://heycam.github.io/webidl/#common-BufferSource) should do the trick, right? I don't think we'd want to enumerate all the various typed array view types.

Re, safelisted mimetype: https://fetch.spec.whatwg.org/#cors-safelisted-request-header - that? 

```
Let <var>corsMode</var> be set to "<code>no-cors</code>" if
object's type is one of <code>BufferSource</code>,
<code>FormData</code>, <code>URLSearchParams</code>, 
<code>USVString</code>, or <code>Blob</code> with a 
<a href="https://fetch.spec.whatwg.org/#cors-safelisted-request-header">CORS-safelisted <var>mimeType</var></a>
value. Otherwise, let <var>corsMode</var> be set to "cors".
```

Getting closer?

**Comment by annevk** [comment]
What would you think about not switching _corsMode_ and just always use "no-cors" and error on erroneous MIME types?

Basically you extract the bytes and MIME type per the same algorithm as in Fetch.

Then you create a Fetch "header" whose "name" is `Content-Type` and value is (the extracted) MIME type. Then you check whether that is a "CORS-safelisted request header". If it's not, throw. Otherwise, proceed.

**Comment by igrigorik** [comment]
That would certainly simplify the spec but it would also disallow `cors`-mode requests via sendBeacon.. If an origin is willing to do the CORS dance, why disallow sendBeacon from supporting that use case?

**Comment by annevk** [comment]
Because the way you go about it is fairly magical and not adding much value. If you want `sendBeacon()` to just be form, I'd rather have it be equally simple.

**Comment by annevk** [comment]
(Even if you wanted to keep "cors", which I no longer think we should do, you can still switch to my approach but rather than fail just switch the mode.)

**Comment by toddreifsteck** [comment]
I believe #33 resolves the block vs. allow list issue. The discussion of how to cleanly spec sendBeacon on top of fetch seems like a future discussion that we should take up in another thread.

**Comment by annevk** [comment]
Well, #33 was resolved with late objections. If you want to move this discussion elsewhere again, be my guest, but it's getting rather frustrating.

And this is not a discussion of how to layer things on top of Fetch, this is a discussion about functionality and utility.

**Comment by igrigorik** [comment]
> Because the way you go about it is fairly magical and not adding much value. If you want sendBeacon() to just be form, I'd rather have it be equally simple.

I don't believe "be like form" is or was our explicit goal. From the discussion on the call earlier this week, the main risk here is that we already have sendBeacon out in the wild and changing it to disallow Blob's runs the risk that we'll break existing deployments. Also, neither Chrome nor Edge has existing telemetry to track which content types are being sent.. so, we'd be making a blind decision. Does FF have something that could help us make the call?

In absence of data to answer the above, I think we can still make progress here as you suggested, but allow cors workflow after checking the content type. Based on your previous comments:

-  Create the following temporary variables:
 - Extract object's byte stream (transmittedData) and MIME type (`mimeType`).
 - Let `headerList` be null.
- If `mimeType` is not null, append a Content-Type header with value mimeType to headerList.
- If `mimeType` is a CORS-safelisted mime type, let `corsMode` be set to "no-cors" and otherwise set to "cors". 

Is that consistent with what you're thinking?

**Comment by annevk** [comment]
Yes, but I would prefer not falling back to "cors" for such an obscure case.

**Comment by igrigorik** [comment]
@annevk does FF have any telemetry on existing content-type usage for sendBeacon?

In absence of real-world usage data my proposal is we adopt the above language that allows "cors" to help normalize behavior between various implementations. As a non-blocking item: add instrumentation to gather telemetry on usage and then revisit the decision to disallow "cors" once we have some numbers.

**Comment by annevk** [comment]
I don’t think we do. Chrome does have CORS telemetry, but might be too generic.

**Comment by igrigorik** [comment]
@annevk thanks. Are you OK with the proposal in my earlier comment, until we get some real-world telemetry?

**Comment by annevk** [comment]
Yeah, modulo some minor wording issues that seems good.

**Comment by igrigorik** [comment]
@annevk @toddreifsteck updated, how does that look?

File: index.html [hunk 1]
```diff
@@ -447,17 +447,26 @@ <h4 id="return-values">Return Value</h4>
               object's byte stream (<var>transmittedData</var>) and MIME type
               (<var>mimeType</var>).
             </li>
-            <li>Let <var>corsMode</var> be set to "<code>cors</code>" if object's type is is <code>Blob</code> and "<code>no-cors</code>" otherwise.</li>
+            <li>Let <var>corsMode</var> be set to "<code>cors</code>".</li>
```
Comment (annevk) [inline index.html]: s/set to//

File: index.html [hunk 2]
```diff
@@ -447,17 +447,26 @@ <h4 id="return-values">Return Value</h4>
...
             <li>Let <var>headerList</var> be null.</li>
           </ul>
         </li>
+        <li>If <var>mimeType</var> is not null:
+          <ul>
+            <li>If <var>mimeType</var> value is a <a href=
+            "https://fetch.spec.whatwg.org/#cors-safelisted-request-header">CORS-safelisted
+            request-header</a> value for <code>Content-Type</code> header, set
```
Comment (annevk) [inline index.html]: the Content-Type*

File: index.html [hunk 3]
```diff
@@ -447,17 +447,26 @@ <h4 id="return-values">Return Value</h4>
...
+          </ul>
+        </li>
         <li>
-          <p>If <var>mimeType</var> is not null, append a
-          <code>Content-Type</code> header with value <var>mimeType</var> to
-          <var>headerList</var>. Append a <code>Accept-Language</code> header
-          with an appropriate <a>value</a> to <var>headerList</var>. Append a
-          <code>Accept</code> header with <code>*/*</code> as the <a>value</a>
-          to <var>headerList</var>.</p>
+          <p>Append a <code>Accept-Language</code> header with an appropriate
```
Comment (annevk) [inline index.html]: an*

File: index.html [hunk 4]
```diff
@@ -447,17 +447,26 @@ <h4 id="return-values">Return Value</h4>
...
+          <a>value</a> to <var>headerList</var>. Append a <code>Accept</code>
+          header with <code>*/*</code> as the <a>value</a> to
+          <var>headerList</var>.</p>
```
Comment (annevk) [inline index.html]: Are Accept-Language/Accept not already done by Fetch in this way?

**Comment by annevk** [comment]
I guess we can land this modulo nits, but we should open an issue to change the language (if needed) based on how https://bugs.chromium.org/p/chromium/issues/detail?id=490015 gets resolved.

File: index.html (hunk 4 above)
Comment (igrigorik) [inline index.html]: Hmm, you're right they are. Guessing, this is a leftover from before we called fetch. I'll remove this.

**Comment by igrigorik** [comment]
@annevk updated, thanks for the help on this one. 

@toddreifsteck can you please review and merge if it all looks good from your end?

File: index.html (hunk 4 above)
Comment (tyoshino) [inline index.html]: Let's mention this in the commit log

**Comment by tyoshino** [comment]
The change looks good.

> Also, neither Chrome nor Edge has existing telemetry to track which content types are being sent.. so, we'd be making a blind decision. Does FF have something that could help us make the call?

I'm adding UMA for http://crbug.com/490015 at https://codereview.chromium.org/2184973002/

> ArrayBuffer

Chrome's sendBeacon has been sending "Content-Type: application/octet-stream" since it was shipped in the middle of 2014 which is not simple (I just noticed it yesterday).

**Comment by igrigorik** [comment]
@tyoshino awesome, curious to see the data!

**Comment by igrigorik** [comment]
> Based on data for only a few days, but we're seeing that ~5% of sendBeacon user pages are using a Blob with non-simple content type for the body to send. -- [tyoshino](https://bugs.chromium.org/p/chromium/issues/detail?id=490015#c40)

That's more than I would have guessed.. I think what we have in this pull is the right solution; I don't think we can drop 'cors' case.

**Comment by plehegar** [comment]
LGTM

**Comment by bzbarsky** [comment]
@plehegar This didn't update https://w3c.github.io/beacon/#privacy to match the normative spec, right?

**Comment by annevk** [comment]
Filed #43 on that section.

**Comment by plehegar** [comment]
@annevk did the right thing to file a new issue.

**Comment by bzbarsky** [comment]
I wasn't sure whether I was just misreading things is all.

**Pull request merged**